src/pages/commits/utils.ts

```diff
@@ -0,0 +1,163 @@
+import { MainlineCommitsQueryVariables } from "gql/generated/types";
+import { TaskStatus } from "types/task";
+import { array } from "utils";
+import { arrayIntersection } from "utils/array";
+
+const { arraySetDifference } = array;
+
+interface filterState {
+  statuses: string[];
+  tasks: string[];
+  variants: string[];
+  requesters: string[];
+}
+interface mainlineCommitOptions {
+  projectID: string;
+  limit: number;
+  skipOrderNumber: number;
+}
+interface commitsPageReducerState {
+  filterState: filterState;
+  mainlineCommitOptions: mainlineCommitOptions;
+}
+
+const getMainlineCommitsQueryVariables = (
+  state: commitsPageReducerState
+): MainlineCommitsQueryVariables => {
+  const variables = {
+    mainlineCommitsOptions: generateMainlineCommitOptionsFromState(state),
+    buildVariantOptions: generateBuildVariantOptionsFromState(state),
+    buildVariantOptionsForGraph: generateBuildVariantOptionsForGraphFromState(
+      state
+    ),
+    buildVariantOptionsForGroupedTasks: generateBuildVariantOptionsForGroupedTasksFromState(
+      state
+    ),
+    buildVariantOptionsForTaskIcons: generateBuildVariantOptionsForTaskIconsFromState(
+      state
+    ),
+  };
+  return variables;
+};
+
+/** getFilterStatus returns an object containing booleans that describe what filters have been applied. */
+const getFilterStatus = (state: filterState) => ({
+  hasFilters:
+    state.requesters.length > 0 ||
+    state.statuses.length > 0 ||
+    state.tasks.length > 0 ||
+    state.variants.length > 0,
+  hasRequesterFilter: state.requesters.length > 0,
+  hasStatuses: state.statuses.length > 0,
+  hasTasks: state.tasks.length > 0,
+  hasVariants: state.variants.length > 0,
+});
+
+const generateBuildVariantOptionsFromState = (
+  state: commitsPageReducerState
+) => {
+  const { filterState } = state;
+  const { statuses, tasks, variants } = filterState;
+
+  const buildVariantOptions = {
+    tasks,
+    variants,
+    statuses,
+  };
+
+  return buildVariantOptions;
+};
+
+const generateBuildVariantOptionsForTaskIconsFromState = (
+  state: commitsPageReducerState
+): MainlineCommitsQueryVariables["buildVariantOptions"] => {
+  const { filterState } = state;
+  const { hasTasks, hasFilters } = getFilterStatus(filterState);
+  const failingStatuses = arrayIntersection(
+    FAILED_STATUSES,
+    filterState.statuses
+  );
+
+  // no filters should show failing icons only
+  // yes filters should show failing icons only if there are failing statuses
+  // yes task filters and status filters should show all icons
+  // yes task filters and no status filters should show all icons
```

Comment:
I think these comments are a little confusing because regardless of whether or not status filters have been applied, as long as there are task filters we should show all icons. It might be better to just say something like `// if task filters applied, then show all icons`